storage/backup/backup-cli/src/coordinators/restore.rs

```diff
@@ -109,89 +106,204 @@ impl RestoreCoordinator {
         )
         .await?;
 
-        let next_txn_version = self
+        let target_version = self.global_opt.target_version;
+        COORDINATOR_TARGET_VERSION.set(target_version as i64);
+
+        // calculate the start_version and replay_version
+        let max_txn_ver = metadata_view
+            .max_transaction_version()?
+            .ok_or_else(|| anyhow!("No transaction backup found."))?;
+        let lhs = self.ledger_history_start_version();
+        let snapshot_before_lhs =
+            metadata_view.select_state_snapshot(std::cmp::min(lhs, max_txn_ver))?;
+
+        let snapshot_before_target = metadata_view
+            .select_state_snapshot(std::cmp::min(self.target_version(), max_txn_ver))?;
+        ensure!(
+            snapshot_before_lhs.is_some() && snapshot_before_target.is_some(),
+            "No snapshot exists before the target version({}) including genesis",
+            target_version
+        );
+        let snapshot_before_lhs = snapshot_before_lhs.unwrap();
+        let snapshot_before_target = snapshot_before_target.unwrap();
+        ensure!(
+            snapshot_before_lhs.version <= snapshot_before_target.version,
+            "snapshot_before_target {} should be larger than or equal to snapshot_before_lhs {}",
+            snapshot_before_target.version,
+            snapshot_before_lhs.version
+        );
+
+        // Two flags for resuming from a previous in-progress restore
+        // Expected version can be used to tell where to resume when applying writesets or replaying txns
+        // Tree restore in progress can be used to tell the status of 2nd snapshot restore
+        let expected_version = self
             .global_opt
             .run_mode
             .get_next_expected_transaction_version()?;
-        if next_txn_version != 0 {
-            // DB is already in workable state
-            info!(
-                next_txn_version = next_txn_version,
-                "DB is ready to accept transactions, start the node to catch up with the chain. \
-                If the node is unable to catch up because the DB is too old, delete the data folder \
-                and bootstrap again.",
-            );
-            return Ok(());
+        let tree_restore_in_progress = self
+            .global_opt
+            .run_mode
+            .get_state_leaf_count(snapshot_before_target.version)
+            > 0;
+
+        info!(
+            lhs = lhs,
+            target_version = target_version,
+            "Starting restore DB from version {} to {}, snapshot_before_lhs: {}, snapshot_before_target: {}, tree restore in progress: {}, expected_version: {} \n\
+            Note: we only guarantee resume from previous in-progress restore. If you want to restore a new DB, please either specify a new target db dir or delete previous in-progress DB in the target db dir.
+            ",
+            lhs,
```

Comment:
hmm, it's mixture of  structural logging ("lhs = lhs") and embedded variables.. shall we use only one style? (In this particular case I thing a big sentence is easier to read)